### PR TITLE
Feature/1672 ngsi arc gis sink

### PR DESCRIPTION
**Pull request by pmo-sdr**
This PR adds support for sending entities to GIS

Issue #1672

### CODE REVIEW

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/DataUtils.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
+/**
+ *
```
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/DataUtils.java]: License header missing (please check in all new files in this PR)

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/CommonConstants.java [hunk 2]
```diff
@@ -91,8 +91,11 @@ private CommonConstants() {
             + "# \n"
             + "# For those usages not covered by the GNU Affero General Public License please "
             + "contact with iot_support at tid dot es";
-    
+    /**
+     *
```
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/CommonConstants.java]: Empty comment?

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/CommonConstants.java [hunk 3]
```diff
@@ -91,8 +91,11 @@ private CommonConstants() {
...
             + "# For those usages not covered by the GNU Affero General Public License please "
             + "contact with iot_support at tid dot es";
-    
+    /**
+     * 
+     *
+     */
     // Enumeration for ManagementInterface
-    public enum LoggingLevels { FATAL, ERROR, WARN, INFO, DEBUG, ALL, OFF }
+    public enum LoggingLevels { FATAL, ERROR, WARN, INFO, DEBUG, TRACE, ALL, OFF }
```
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/CommonConstants.java]: Is this new TRACE level really necessary? Why don't reuse DEBUG for it?

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/DataUtils.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * @author PMO Santander Smart City - Ayuntamiento de Santander
+ *
+ */
+public final class DataUtils {
+
+    private static final String[] PATTERNS = {
```
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/DataUtils.java]: A brief comment above this line should explain the pattern. It is not evident just looking it :)

File: cygnus-ngsi/pom.xml [hunk 5]
```diff
@@ -1,165 +1,204 @@
 <project xmlns="http://maven.apache.org/POM/4.0.0"
```
Comment (fgalan) [inline cygnus-ngsi/pom.xml]: It seems that some pretty printer has ruineed the whole pom.xml :(

Could you come back to the original one in master adding just the elements needed by the new sink so the PR can clearly show the modifications?

File: cygnus-ngsi/src/main/assembly/assemblyCompile.xml [hunk 6]
```diff
@@ -0,0 +1,24 @@
+<assembly
```
Comment (fgalan) [inline cygnus-ngsi/src/main/assembly/assemblyCompile.xml]: Could you elaborate on the purpose of this file, please?

**Comment by fgalan** [comment]
This file has been added to the PR:

cygnus-ngsi/src/main/resources/ArcgisUtils-0.0.1-SNAPSHOT.jar 

Is this intentional? .jar files are not typically included in code repositories (the right place would be a Maven repository and downloaded from it at building time)

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/utils/NGSIUtilsForTests.java [hunk 7]
```diff
@@ -94,7 +96,24 @@ public static Context createContextForOrion(String orionHost, String orionPort,
...
+
+    /**
+     * Creates a Flume context for Arcgis sinks.
+     * 
+     * @param url
+     * @param username
+     * @param password
+     * @return
+     */
+    public static Context createContextForArcGis(String url, String username,
```
Comment (fgalan) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/utils/NGSIUtilsForTests.java]: PR should include also a unit test to cover the new functionality. I see this addition to NGSIUtilsForTest so I guess that was your idea, but the PR doesn't include any test at the end...

**Comment by fgalan** [comment]
Thank you very much for your contribution! Please find a first round of feedback in above comments.

File: docker/cygnus-ngsi/agent.conf [hunk 8]
```diff
@@ -221,6 +221,21 @@ cygnus-ngsi.sinks.elasticsearch-sink.ssl = false
 #cygnus-ngsi.sinks.elasticsearch-sink.cache_flash_interval_sec = 0
 
 
+# NGSIArcGisSink configuration
+cygnus-ngsi.sinks.arcgis-sink.type = com.telefonica.iot.cygnus.sinks.NGSIArcGisSink
+cygnus-ngsi.sinks.arcgis-sink.channel = arcgis-channel-santander
```
Comment (AlvaroVega) [inline docker/cygnus-ngsi/agent.conf]: arcgis-channel-santander -> arcgis-channel ?

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/DataUtils.java (hunk 1 above)
Comment (pmo-sdr) [reply cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/DataUtils.java]: Done

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/CommonConstants.java (hunk 2 above)
Comment (pmo-sdr) [reply cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/CommonConstants.java]: Removed

File: cygnus-ngsi/pom.xml (hunk 5 above)
Comment (pmo-sdr) [reply cygnus-ngsi/pom.xml]: I changed the pom.xml and I added new library and repository.

File: docker/cygnus-ngsi/agent.conf (hunk 8 above)
Comment (pmo-sdr) [reply docker/cygnus-ngsi/agent.conf]: Yes, I changed it

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/DataUtils.java (hunk 4 above)
Comment (pmo-sdr) [reply cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/DataUtils.java]: I commented the pattern where I use different regular expressions

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/CommonConstants.java (hunk 3 above)
Comment (pmo-sdr) [reply cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/CommonConstants.java]: Done

File: cygnus-ngsi/src/main/assembly/assemblyCompile.xml (hunk 6 above)
Comment (pmo-sdr) [reply cygnus-ngsi/src/main/assembly/assemblyCompile.xml]: This file is a assembly which we use to add the ArcgisUtils-0.0.1-SNAPSHOT.jar to cygnus-ngsi.

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/utils/NGSIUtilsForTests.java (hunk 7 above)
Comment (pmo-sdr) [reply cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/utils/NGSIUtilsForTests.java]: Added NGSIArcGisSinkTest

File: cygnus-ngsi/README.md [hunk 9]
```diff
@@ -284,6 +285,7 @@ Detailed information regarding cygus-ngsi can be found in the [Installation and
   <tr><td>Raw-snapsot analysis</td><td>1.6.0</td></tr>
   <tr><td>NGSIOrionSink</td><td>First implementation</td><td>1.10.0</td></tr>
   <tr><td>NGSIElasticsearchSink</td><td>First implementation</td><td>1.15.0</td></tr>
+  <tr><td>NGSIArcGisSink</td><td>First implementation</td><td>1.15.0</td></tr>
```
Comment (fgalan) [inline cygnus-ngsi/README.md]: Cygnus 1.15.0 window has passed (it will be released today). Thus, please change this to 1.16.0.

Comment (pmo-sdr) [reply cygnus-ngsi/README.md]: Done

File: docker/cygnus-ngsi/cygnus-entrypoint.sh [hunk 10]
```diff
@@ -953,6 +953,74 @@ elif [ "$CYGNUS_ELASTICSEARCH_HOST" != "" ]; then
...
+            if [ "${CYGNUS_ARCGIS_SKIP_NAME_MAPPINGS_GENERATION,,}" != "true" ]; then
+                cp -p ${FLUME_HOME}/conf/name_mappings.conf ${FLUME_HOME}/conf/${NAMEMAPPING_CONF_FILE}
+            fi
+            sed -i '/'${CYGNUS_AGENT_NAME}'.sources.http-source.interceptors.nmi.name_mappings_conf_file/c '${CYGNUS_AGENT_NAME}'.sources.http-source.interceptors.nmi.name_mappings_conf_file = '${FLUME_HOME}/conf/${NAMEMAPPING_CONF_FILE} ${FLUME_HOME}/conf/${AGENT_CONF_FILE}
+            sed -i '/'${CYGNUS_AGENT_NAME}'.sources.http-source.interceptors =/c '${CYGNUS_AGENT_NAME}'.sources.http-source.interceptors = ts nmi' ${FLUME_HOME}/conf/${AGENT_CONF_FILE}
+        fi
+    fi
+    sed -i 's/'${CYGNUS_AGENT_NAME}'.sinks =/'${CYGNUS_AGENT_NAME}'.sinks = arcgis-sink /g' ${FLUME_HOME}/conf/${AGENT_CONF_FILE}
+    sed -i 's/'${CYGNUS_AGENT_NAME}'.channels =/'${CYGNUS_AGENT_NAME}'.channels = arcgis-channel /g' ${FLUME_HOME}/conf/${AGENT_CONF_FILE}
+    sed -i '/'${CYGNUS_AGENT_NAME}'.sources.http-source.port/c '${CYGNUS_AGENT_NAME}'.sources.http-source.port = '5050 ${FLUME_HOME}/conf/${AGENT_CONF_FILE}
```
Comment (AlvaroVega) [inline docker/cygnus-ngsi/cygnus-entrypoint.sh]: 5050 is already port for mysql when mutiagent option is enabled. Please, set the next free port (5059) according with: https://github.com/telefonicaid/fiware-cygnus/blob/master/doc/cygnus-ngsi/installation_and_administration_guide/install_with_docker.md#using-a-specific-configuration

Comment (AlvaroVega) [reply docker/cygnus-ngsi/cygnus-entrypoint.sh]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1683/commits/4e10a96bef8248565cad79b9b4fbfcadfc5c21b5

**Pull request merged**